tokio/src/time/driver/entry.rs

```diff
@@ -268,6 +268,7 @@ impl Entry {
 
         let mut curr = entry.state.load(SeqCst);
         let mut notify;
+        let mut wake_task;
```

Comment:
It's a little unfortunate another variable needs to be introduced to track this, but due to the rest of the implementation it's a little tough to avoid it.. I was thinking maybe we could just check after the loop, but since `next` is scoped to the loop we can't really check if the new value is elapsed without another atomic load.

Reply:
Note that the value of `next` only depends on `when` and `elapsed`.  Both of those are immutable so It would be possible to calculate `next` before the loop and decide whether to wake the waker and queue the entry by inspecting `curr` and `next` after the loop:

```Rust
        let next = if when <= elapsed {
            ELAPSED
        } else {
            when
        };

        let mut curr = entry.state.load(SeqCst);

        loop {
            // In these two cases, there is no work to do when resetting the
            // timer. If the `Entry` is in an error state, then it cannot be
            // used anymore. If resetting the entry to the current value, then
            // the reset is a noop.
            if curr == ERROR || curr == when {
                return;
            }

            let actual = entry.state.compare_and_swap(curr, next, SeqCst);

            if curr == actual {
                break;
            }

            curr = actual;
        }

        if !is_elapsed(curr) && is_elapsed(next) {
            entry.waker.wake();
        }

        if !is_elapsed(curr) || !is_elapsed(next) {
            let _ = inner.queue(entry);
        }
```

I think I prefer this; it keeps the core loop smaller and separates out the logic for deciding what to do based on the state change.  Do you agree?

Reply:
I'd be good with this. I think the two if statements at the bottom could use comments explaining the conditions and why the waker may need to be woken or the entry queued, but I agree the loop does get a little easier to follow.

Reply:
I've made this change and added the requested comments.